src/magicbox/ResultPreviewsGrid.ts

```diff
@@ -0,0 +1,385 @@
+
+export type ReceivedSearchResultPreviews = ReceivedSearchResultPreview[];
+
+/**
+ * This class renders a grid of result previews from [`QuerySuggestPreview`]{@link QuerySuggestPreview} inside a given container and allows navigation within it.
+ * It waits to receive a first [`SearchResultPreview`]{@link SearchResultPreview} before creating any HTML element.
+ */
+export class ResultPreviewsGrid {
+  public static ContainerClassName = 'coveo-preview-container';
+  public static HeaderClassName = 'coveo-preview-header';
+  public static ResultsContainerClassName = 'coveo-preview-results';
+
+  private static setPreviewId(element: HTMLElement, id: number) {
+    element.dataset.previewId = id.toString();
+  }
+
+  private static getPreviewId(element: HTMLElement) {
+    return element.dataset.previewId ? parseInt(element.dataset.previewId, 10) : null;
+  }
+
+  private resultContainerElements?: {
+    container: Dom;
+    status: Dom;
+    results: Dom;
+  };
+  private options: IResultPreviewsGridOptions;
+  private queryProcessingRejector: Function;
+  private activePreviews: ActiveSearchResultPreview[];
+  private keyboardSelectionMode: boolean;
+
+  constructor(private parentContainer: HTMLElement, options: IResultPreviewsGridOptions = {}) {
+    this.options = defaults(options, <IResultPreviewsGridOptions>{
+      timeout: 2000,
+      maximumPreviews: 6,
+      selectedClass: 'magic-box-selected'
+    });
+  }
+
+  private get previewsPerRow() {
+    // To account for every CSS that may span previews over multiple rows, this solution was found: https://stackoverflow.com/a/49090306
+    if (this.activePreviews.length === 0) {
+      return null;
+    }
+    const firstVerticalOffset = this.activePreviews[0].dom.offsetTop;
+    const firstIndexOnNextRow = _.findIndex(this.activePreviews, preview => preview.dom.offsetTop !== firstVerticalOffset);
+    return firstIndexOnNextRow !== -1 ? firstIndexOnNextRow : this.activePreviews.length;
+  }
+
+  public receiveLegacyOptions(options: IResultPreviewGridLegacyOptions) {
+    Object.keys(this.options).forEach(optionName => (this.options[optionName] = options[optionName] || this.options[optionName]));
+  }
+
+  /**
+   * Waits for one of the following conditions to be true then creates a container and fills it with the results.
+   * 1. No query was given
+   * 2. All queries were completed
+   * 3. Enough queries were completed to fill the maximum previews
+   * 4. [`timeout`]{@link ResultPreviewsGrid.options.timeout} has passed
+   * If the function is called again while it's still processing, the previous call is cancelled and overriden by the new one.
+   */
+  public receiveSearchResultPreviews(
+    variantSearchResultPreviewsQueries: Array<Promise<ReceivedSearchResultPreviews> | ReceivedSearchResultPreviews>,
+    getCompletionMessage?: ((activePreviews: SearchResultPreview[]) => string) | void
+  ): Promise<SearchResultPreview[]> {
+    return new Promise((resolve, reject) => {
```

Comment:
I wrote this because I needed to resolve the function from inner functions (e.g., `showAndReturn`), but I'm sure there would be a way to use `async/await` instead and I'll change it.